pxr/imaging/hgiMetal/shaderGenerator.mm

```diff
@@ -303,9 +303,9 @@ void _Init(
         " atomic_fetch_add_explicit(&a, v, memory_order_relaxed)\n"
         "#define ATOMIC_EXCHANGE(a, desired)"
         " atomic_exchange_explicit(&a, desired, memory_order_relaxed)\n"
-        "#define ATOMIC_COMP_SWAP(a, expected, desired) \\"
-        "  atomic_compare_exchange_strong_explicit(&a, &expected, desired, "
-        "memory_order_relaxed, memory_order_relaxed)\n"
+        "#define ATOMIC_COMP_SWAP(a, expected, desired)"
+        "  atomic_compare_exchange_weak_explicit(&a, &expected, desired, "
+        "memory_order_relaxed, memory_order_relaxed) ? expected : expected;\n"
```

Comment:
Thanks for fixing the return value here!

A few issues with this function:
- This predates this PR, but I'm a bit worried about this macro introducing side effects (namely that expected is an out value) that aren't present in GL or Vulkan.
- I'm unclear if the comparable GLSL semantics are weak or strong.  But if weak works like I expect, we need to handle spurious failures in client code and there's a subtle/minor bug in your deep picking shader, which will likely show up in other CAS uses as well.

Reply:
Note: GLSL has weak semantics too; so there's still some care needed, and I think there's still a latent bug in your usage of it, but as far as this file is concerned the only issue is the side effect.  Maybe we can make a scoped temporary copy of expected?

Reply:
Having a very hard time pinning this down, but based on specs and other projects cross-compiling GLSL to MSL, I think GLSL CAS is strong, and MSL is weak.  I have an alternate implementation of the macro for Metal that I'm going to test, to give it strong semantics and no side effects.